File: Data Structures & Algorithms/count-good-nodes-in-binary-tree/submission-2.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */

class Solution {
public:
    int goodNodes(TreeNode* root) {
        return dfs(root, INT_MIN);
        
    }
    int dfs(TreeNode *root, int top) {
        if (root == nullptr) return 0;
        int res = root->val >= top ? 1 : 0;
        top = max(root->val, top);
        res+= dfs(root->left, top);
        res+= dfs(root->right, top);
        return res;
    }
};

File: Data Structures & Algorithms/interleaving-string/submission-2.py
class Solution:
    def isInterleave(self, s1: str, s2: str, s3: str) -> bool:
        if len(s1) + len(s2) != len(s3):
            return False
        memo = {}
        def dfs(i, j):
            if i == len(s1) and j == len(s2):
                return True
            key = (i, j)
            if key in memo:
                return memo[key]
            res = False
            if i < len(s1) and s1[i] == s3[i + j]:
                res = dfs(i + 1, j) or res
            if j < len(s2) and s2[j] == s3[i + j]:
                res = dfs(i, j + 1) or res
            memo[key] = res
            return memo[key]
        return dfs(0, 0)